mem/mem.go

```diff
@@ -0,0 +1,325 @@
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package mem
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"sync/atomic"
+)
+
+// DefaultChannelSize is the default Pool channel size.
+//
+// See the WithChannelSize function for more details.
+const DefaultChannelSize uint16 = 4
+
+// DefaultSegListSizes are the default seg list sizes.
+//
+// See the WithSegListSizes function for more details.
+var DefaultSegListSizes = []int{
+	8,
+	16,
+	64,
+	256,
+	4096,
+	65536,
+	1048576,
+	16777216,
+}
+
+var globalPool = NewPool(WithNoLocking(), WithDisabled())
+
+// Global returns the global Pool instance.
+//
+// It is generally preferable to create a Pool instance for your application and
+// pass it through to the places it is needed, however this will prove difficult for
+// existing applications wishing to migrate to the Pool API.
+//
+// The Global Pool is disabled by default. To enable pooling in your application using
+// the Get* generated functions for each message, call Global().Enable() at initialization.
+func Global() *Pool {
+	return globalPool
+}
+
+// Bytes represents a byte slice created from a Pool.
+//
+// Only create these from Pools.
+type Bytes struct {
+	segListPool *segListPool
+	value       []byte
+	valueLen    int
+}
+
+// NewUnmanagedBytes creates a new Bytes that is not backed by any Pool.
+//
+// Users should generally not call this directly.
+func NewUnmanagedBytes(valueLen int) *Bytes {
+	return &Bytes{
+		value:    make([]byte, valueLen),
+		valueLen: valueLen,
+	}
+}
+
+// Value gets the value.
+func (b *Bytes) Value() []byte {
+	return b.value[:b.valueLen]
+}
+
+// Len gets the length.
+//
+// This is equivalent to len(b.Value()).
+func (b *Bytes) Len() int {
+	return b.valueLen
+}
+
+// Truncate truncates the value to the given length.
+func (b *Bytes) Truncate(valueLen int) {
+	b.valueLen = valueLen
+}
+
+// MemsetZero sets the data inside the value to zero.
+//
+// Optimized per https://golang.org/cl/137880043
+func (b *Bytes) MemsetZero() {
+	for i := range b.value {
+		b.value[i] = 0
+	}
+}
+
+// Recycle recycles the Bytes.
+//
+// If this is nil, this is a no-op.
+func (b *Bytes) Recycle() {
+	if b == nil {
+		return
+	}
+	// note this is possible if we did not have a big enough seg list
+	if b.segListPool != nil {
+		b.segListPool.put(b)
+	}
+}
+
+// PooledMessage is a message that can be pooled.
+//
+// This generally is meant for proto.Messages but is compatible with
+// anything that has a Reset and Recycle method.
+type PooledMessage interface {
+	Reset()
+	Recycle()
+}
+
+// Pool is a pool of PooledMessages and Bytes.
+type Pool struct {
```

Comment:
Document explicitly if methods on the pool are generally thread safe (I get the impression they are?).